lib/ui/channel_buffers.dart

```diff
@@ -0,0 +1,83 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of dart.ui;
+
+/// A saved platform message for a channel with its callback.
+class StoredMessage {
+  final ByteData _data;
+  final PlatformMessageResponseCallback _callback;
+
+  StoredMessage(this._data, this._callback);
+  ByteData get data => _data;
+  PlatformMessageResponseCallback get callback => _callback;
+}
+
+/// A fixed-size circular queue.
+class RingBuffer<T> {
+  final collection.ListQueue<T> _queue;
+  final int _capacity;
+
+  RingBuffer(this._capacity) 
+    : _queue = collection.ListQueue<T>(_capacity);
+
+  int get length => _queue.length;
+
+  int get capacity => _capacity;
+
+  bool get isEmpty => _queue.isEmpty;
+
+  /// Returns true on overflow.
+  bool push(T val) {
+    bool overflow = false;
+    while (_queue.length >= _capacity) {
+      _queue.removeFirst();
+      overflow = true;
+    }
+    _queue.addLast(val);
+    return overflow;
+  }
+
+  /// Returns null when empty.
+  T pop() {
+    return _queue.isEmpty ? null : _queue.removeFirst();
+  }
+}
+
+/// Storage of channel messages until the channels are completely routed.
+class ChannelBuffers {
```

Comment:
maybe `PlatformChannelBuffer`?

The doc string here should have more information, paritcularly about when push/pop might not behave as expected and what a developer should do about it in those case.

Reply:
Updated the docstring.

I prefer ChannelBuffers, the reason is that what clients know about are things like MethodChannel.   If they are named PlatformChannelBuffers and MethodChannel, it seems to hint they aren't related.  ChannelBuffers and MethodChannel sounds plausible that they are related.